### PR TITLE
Assert that ActionListener.runBefore/runAfter are executed once

### PR DESCRIPTION
**Pull request by tlrx**
I've recently encountered a bug (illustrated by a test in this pull request) where a `ActionListener.runAfter` listener got called multiple times because of an unexpected exception was thrown in a wrapped listener.

Asserting that the `runAfter` or `releaseAfter` is called only once would have helped a lot to catch the bug. We don't enforce listeners (or delegating listeners) to be called once and some of them are designed on purpose to be called multiple times, yet I think we can expect `runBefore/runAfter` to be used to release resources and as such they should be run once. 

I remain open to contrary opinions of course.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -402,10 +405,19 @@ public void onFailure(Exception e) {
             try {
                 super.onFailure(e);
             } finally {
+                assert assertNotExecuted();
                 runAfter.run();
             }
         }
 
+        private boolean assertNotExecuted() {
+            if (Assertions.ENABLED) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:414]: We can't get here if assertions are disabled I think

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 2]
```diff
@@ -452,6 +467,14 @@ public void onFailure(Exception e) {
             super.onFailure(e);
         }
 
+        private boolean assertNotExecuted() {
+            if (Assertions.ENABLED) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:471]: We can't get here if assertions are disabled I think

File: server/src/test/java/org/elasticsearch/action/ActionListenerTests.java [hunk 3]
```diff
@@ -454,6 +457,63 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalStateException.class));
     }
 
+    public void testRunBeforeThrowsAssertionErrorIfExecutedMoreThanOnce() {
+        assumeTrue("test only works with assertions enabled", Assertions.ENABLED);
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:461]: Huh TIL it's possible to run tests with assertions disabled! I wouldn't be surprised if this didn't work in lots of other ways too tho.

**Review by DaveCTurner — COMMENTED** [review]
LGTM

**Review by DaveCTurner — APPROVED** [review]
Sorry forgot to press the button.

FWIW I think the way `ActionListener#wrap` sometimes calls both `onResponse` and `onFailure` is kinda evil. Useful sometimes, but evil nonetheless.

File: server/src/main/java/org/elasticsearch/action/ActionListener.java (hunk 2 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/action/ActionListener.java]: Right, thanks! e62d927eef846b6337eec8a7272957c75a076bb2

File: server/src/test/java/org/elasticsearch/action/ActionListenerTests.java (hunk 3 above)
Comment (tlrx) [reply server/src/test/java/org/elasticsearch/action/ActionListenerTests.java:461]: Yep...

**Comment by tlrx** [comment]
@elasticmachine run update branch

**Comment by tlrx** [comment]
@elasticmachine update branch

**Pull request merged**